rep-2017.rst

```diff
@@ -0,0 +1,359 @@
+
+The purpose of this REP is to propose a feature to the ros client library that allows it to accept thread attributes, store them internally, and provide interfaces to pass them to upper language bindings (e.g., rclcpp, rclpy, or future ones).  By incorporating this feature and enabling language bindings to read the thread attributes, ROS 2 can effectively adopt them for real-time systems or environments that require strict CPU resource optimization, eliminating the need for code modifications in each specific environment.
+
+Specification
+=============
+
+This chapter provides the interface specifications depicted as [Parameter configuration] and [Language binding] below.
+
+::
+
+                              +-----------+ +------------+
+                              |   rclcpp  | |   rclpy    | ...
+                              +-----^-----+ +-----^------+
+                                    |             |
+                               ...[Language bindgings]...
+                                    |             |
+                              +-----+-------------+-----------+
+    ------Env. variable ------>                               |
+    [Parameter configuration] |             rcl               |
+    -- Cmd line parameters --->                               |
+                              |                               |
+                              +-------------------------------+
+
+
+Parameter configuration
+-----------------------
+
+This interface enables robot system integrators to pass a set of thread attribute parameters to the rcl, which internally stores them.  When the language bindings (e.g., rclcpp, rclpy, and so on) need to set their executor's thread attributes, the rcl will provide these parameters.
+
+The interface offers two methods to pass these parameters: through environment variables and command line parameters.  The environment variable option is suitable for system-wide attribute settings, while the command line parameter option is designed for specific process configurations.  Therefore, the environment variable method has a lower priority, and rcl ignores it if parameters are provided via the command line parameter method.
+
+The thread attribute parameters consist of these items for each thread, as below.
+
+* Thread name
+* Core affinity
+* Scheduling policy
+* Priority
+
+When using the parameter passing feature, users should create an array containing sets of these items as thread attribute parameters for the thread pool used by ROS 2 executors.  The text format that users should comply with for describing parameters is YAML.  Users can choose to pass the parameters either as a string or from a file.
```

Comment:
There might be multiple executors running in the process space. in that case, how does this behave?

Reply:
Even if the process has multiple executors, a user application can be implemented to choose appropriate ones for each executor, as I mentioned earlier.

Reply:
What if the process has multiple single threaded executors? Does your proposed solution described in https://github.com/ros-infrastructure/rep/pull/385#discussion_r1347295686 cover this case?

It would be helpful to see some usage examples with different scenarios. That would help to understand how the proposed solution works top to bottom.

Reply:
@carlossvg 
> What if the process has multiple single threaded executors? Does your proposed solution described in https://github.com/ros-infrastructure/rep/pull/385#discussion_r1347295686 cover this case?

Yes. In such cases, as described below, the user application retrieves a thread attribute for each single thread executor and then passes it to the appropriate single thread executor one by one.
````
[Applications]

Applications retrieve thread attributes from rcl and pass one of them to the executor's constructor. The executor then configures the thread(s) accordingly.
````
> It would be helpful to see some usage examples with different scenarios. That would help to understand how the proposed solution works top to bottom.

Our team is going to update the [pull-request](https://github.com/ros2/rclcpp/pull/2205) to demonstrate how applications pass the thread attributes to the existing executors (single thread/multit hread) and create a tutorial to show how we can use the infrastructure by using an existing sample application like below.
https://github.com/ros-realtime/reference-system

Reply:
@carlossvg   
> What if the process has multiple single threaded executors? Does your proposed solution described in [#385 (comment)](https://github.com/ros-infrastructure/rep/pull/385#discussion_r1347295686) cover this case?
> 
> It would be helpful to see some usage examples with different scenarios. That would help to understand how the proposed solution works top to bottom.

Sorry for my late update.
The usage example explaining that our proposal is able to cover the case where a ROS 2 process has multiple single-threaded executors is prepared, as below.

[Getting started]
https://github.com/esol-community/reference-system/blob/rep2017_demo/misc/getting_started_thread_attribute_configuration_extention_rep2017.md

[code]
https://github.com/esol-community/reference-system/blob/rep2017_demo/autoware_reference_system/src/ros2/executor/autoware_default_prioritized_using_rep2017.cpp

Seeing the code above, you will find there are two methods to use the thread attributes. 

The first one is to retrieve them by user application code (USE_USER_IMPLEMENTED_ATTR_FIND_ROUTINE). In this method, the user application code retrieves the thread attributes and passes them to the executor via a parameter for its constructor. Then, the executor sets the passed attribute to the thread it creates.

The second is to delegate the retrieving procedure to executors by passing some hints as the constructor's parameters. In this method, given a hint (like RCLCPP_EXECUTOR_HOTPATH), the executor searches the dedicated thread attribute matching it and sets the attribute to the thread created by the executors.

Reply:
@carlossvg 
Thanks for your feedback.

> I found a bit confusing how the default thread attributes are set for executors. This is because there is no one to one mapping between threads, executors and thread options.

Does your concern relate to the situation where there are no thread attributes being passed, either via a command line parameter or via an environment variable?
If so, the default behavior is the same as the existing one. In other words, the thread attributes for the caller's thread will be inherited by the called executor(s). 

> name field in rcutils_thread_attr_s struct refers to a thread name. However, this seems to me more like the name of a thread attributes configuration. This configuration can be reused in different thread or executor.

It is a good point indeed. My intention is that the field is used as a hint (or hints) for the retrieving process in applications or executors. Considering your comment, "tag" is more suitable than "name" for the purpose.

```
name:              String value representing the name of the thread
->
tag:               String value used as a hint (or hints) to retrieve an appropriate parameter set in application code or executors.
```
Does the above go along with your opinion?

Reply:
>Does your concern relate to the situation where there are no thread attributes being passed, either via a command line parameter or via an environment variable?

Not really, the default behavior is fine for me. My comment was related to the hint/tag comment.

>Does the above go along with your opinion?

Yes, I find this name more suitable. Thanks.